Data/pubtator_to_xml.py

```diff
@@ -0,0 +1,215 @@
+    """
+    article = {}
+
+    # title
+    title_heading = file_lines[0].split('|')
+    title_len = len(title_heading[2])
+    article["Document ID"] = title_heading[0]
+    article["Title"] = title_heading[2]
+    title_len = len(title_heading[2])
+    # abstract
+    abstract_heading = file_lines[1].split("|")
+    article["Abstract"] = abstract_heading[2]
+
+    # set up the csv reader
+    annts = csv.DictReader(file_lines[2:], fieldnames=['Document', 'Start', 'End', 'Term', 'Type', 'ID'], delimiter=str("\t"))
+    sorted_annts = sorted(annts, key=lambda x: x["Start"])
+    article["Title_Annot"] = filter(lambda x: x["Start"] < title_len, sorted_annts)
+    article["Abstract_Annot"] = filter(lambda x: x["Start"] > title_len, sorted_annts)
+
+    return article
+
+
+def bioconcepts2pubtator_offsets(input_file):
+    """Bioconcepts to pubtator
+
+    Yields an article that is a dictionary described in the article generator
+    function.
+
+    Keywords:
+    input_file - the name of the bioconcepts2putator_offset file (obtained from pubtator's ftp site: ftp://ftp.ncbi.nlm.nih.gov/pub/lu/PubTator/)
+    """
+    file_lines = list()
+    if ".gz" in input_file:
+        f = gzip.open(input_file, "rb")
+    else:
+        f = open(input_file, "rb")
+
+    for line in f:
+        # Convert "illegal chracters" (i.e. < > &) in the main text
+        # into html entities
+        line = line.rstrip()
+        if line:
+            file_lines.append(line)
+        else:
+            yield pubtator_stanza_to_article(file_lines)
+            file_lines = list()
+
+    # we missed a document because the file didn't
+    # end in a new line
+    if len(file_lines) > 0:
+        yield pubtator_stanza_to_article(file_lines)
+
+    f.close()
+
+
+def convert_pubtator(input_file, output_file=None):
+    """Convert pubtators annotation list to BioC XML
+
+    Keyword Arguments:
+    input_file -- the path of pubtators annotation file
+    output_file -- the path to output the converted text
+    """
+    if output_file is None:
+        output_file = "bioc-converted-docs.xml"
+
+    # Set up BioCWriter to write specifically Pubtator
+    # Can change to incorporate other sources besides pubtator
+    writer = BioCWriter()
+    writer.collection = BioCCollection()
+    collection = writer.collection
+    collection.date = time.strftime("%Y/%m/%d")
+    collection.source = "Pubtator"
+    collection.key = "Pubtator.key"
+
+    with open(output_file, 'wb') as g:
+
+        # Have to manually do this because hangs otherwise
+        # Write the head of the xml file
+        xml_header = writer.tostring('UTF-8')
+        xml_tail = '</collection>\n'
+        xml_head = xml_header[:-len(xml_tail)]
+        g.write(xml_head)
+
+        article_generator = bioconcepts2pubtator_offsets(input_file)
+        # Write each article in BioC format
+        for article in tqdm.tqdm(article_generator):
+            id_index = 0
```

Comment:
Let's move `id_index` to right above the for loops where it's used. Also `id_index = id_index + 1` is the same as `id_index += 1` if you want the brevity.